### PR TITLE
LPS-150279 Enable predefined values to the FDS filters

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-150279 Enable predefined values to the Frontend Data Set filters](https://issues.liferay.com/browse/LPS-150279)
- [Parent Epic](https://issues.liferay.com/browse/LPS-150047)

### What is the goal of this PR?

In this PR, we are enabling preloading of data into all FDS filters. To achieve this, we are both extending filter object definition in the backend, and handling in the frontend.

@beltranrengifo @javierdearcos @gabrielwas @dsanz 

Notes:
- We are using a rather generic `Map<String, Object>` as a data structure for preloaded data. This is because different filters have different settable elements. For example, `CheckboxesFilter` has the `exclude` toggle, while `DateRangeFilter` has min and max date objects. This kind of API will support any future filters we add, as well as potential custom filters. The only requirement will be that the data structure in `Object` is serializable. I hope that this fully provides the API that the Sierra team requires.
- We are adding a sample for checkbox filter in the sample module. In future tasks, we plan to add samples of all filters, with and without preloading.
- We are removing `NumbersFilter` and `TextFilter`, as they are unused.
- I refactored the way filters are handled in FDS frontend. Up to this PR, there were two layers of filters state, one in the top components, and another in `FiltersContext`. After this PR, filters are only in one state, in the top component. It is passed down  as a part of `DataSetContext`. The state needs to be on top level, since preloaded filters immediately influence data loading.
- We are standardizing a JS 'interface' that any filter implementation should follow: `getOdataString`, `getSelectedItemsLabel` and the component itself. In theory these could be extracted to `Filter` or to some util, but I like that all the logic of the specific filter is contained in one file.
- I added the explicit `active` flag for filter objects. This is mainly for code readability. Currently we are check for filter value.
- I refactored all functions the code touches to have a single object argument.
- Renames:
  - `value` / `valueProp` --> `selectedData`
  - `formattedValue` --> `selectedItemsLabel`
  - `formatValue` --> `getSelectedItemsLabel`
  - `updateFilterState` --> `setFilter`
  - filters/index.js --> filters/Filter.js

### What does it look like?

after

#### Testing preloading

1. Place sample FDS module on the page. See gif above.

#### Testing possible regressions in other filters

- Setup minium site for commerce. Make an order. There are a lot of steps for this, let me know if I need to go in details.
- `DateRangeFilter` - orders > order date range 
- `AutocompleteFilter` - orders > account
- `RadioFilter` - products > product type

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/filter/FDSFilter.java [hunk 1]
```diff
@@ -23,6 +25,8 @@
 
 	public String getLabel();
 
+	public Map<String, Object> getPreloadedData();
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/filter/FDSFilter.java:28]: If we use a `default` method here:
```
public default Map<String, Object> getPreloadedData() {
     return null;
}
```
then we would not need to override it in the Base classes, only just in those that need to return anything. Also, we'd not force other future Base classes to implement it unless needed

I don't remember well how this affects semver, we shall take a look to how this is applied in other interfaces

**Review by beltranrengifo — COMMENTED** [review]
Hi @markocikos !
Everything looks good. Liked the Filter component ❤️ 
Not much to say on my side about the backend part. Regarding the frontend, I left you some minor comments, feel free to proceed as you may consider.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/DataSet.js [hunk 2]
```diff
@@ -621,6 +664,8 @@ const DataSet = ({
 				selectedItemsKey,
 				selectedItemsValue,
 				selectionType,
+				setFilter,
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/DataSet.js:667]: Yeah, I'm not happy with this either. I though of many places to put this convenience function, but none of them fits perfectly:
- as a util (not a good separation of concerns)
- in context (requires state in context, too many changes for no real benefit)
- in top component, like in this PR (awkward props)
- in leaf components (lots of duplication)

I'll try to make some sort of a compromise solution.

**Comment by markocikos** [comment]
Reviews green, CI green...

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**